sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -499,6 +506,9 @@ case class Cast(
   def this(child: Expression, dataType: DataType, timeZoneId: Option[String]) =
     this(child, dataType, timeZoneId, evalMode = EvalMode.fromSQLConf(SQLConf.get))
 
+  def this(child: Expression, dt: Expression) =
```

Comment:
This is the probably most interesting part of this change. Here, we simply use the output type of the expression to determine what the target type of the type cast is.

There is a similar piece of code in the `from_json` function which takes an expression as the second parameter for the schema, but evaluates it to string and then parses the string into the type.